jl4-core/src/L4/TypeCheck/Types.hs

```diff
@@ -583,18 +584,28 @@ prune m = do
 
       -- We have a success, we don't want a second one
       procPlain a []                   = [a]
-      procPlain a ((Plain _, _)  : []) = [first (With (MkCheckErrorWithContext InternalAmbiguityError ctx)) a]
-      procPlain _ ((Plain _, _)  : cs) = [last cs]
       procPlain a ((With _ _, _) : cs) = procPlain a cs
+      procPlain a cs
+        | plain = [first (With (MkCheckErrorWithContext InternalAmbiguityError ctx)) a]
+        -- NOTE: in the case of ambiguity errors, the ambiguity error will always occurs as
+        -- the last element in the list
+        | otherwise = [last cs]
+        where
+          plain = allPlain $ map fst cs
```

Comment:
Is this correct?
Wouldn't that mean that '[Plain, Plain, With]' reports 'With', even though it should be ambiguous?

Reply:
It would report With but then the With would contain an ambiguity error. That's exactly the assumption that we're (ab)using